### ISSUE TITLE
Squash rebase merges please for easier release tracking

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#2877]
Each release is showing 20-30 commits, intermingled by time, some going back way in history. It makes it very hard to know what was merged with which releases exactly. It's also difficult to view the source code history, commits are intermingled by time rather than grouped by merge. Third, it's more difficult to revert a merge commit compared to a single rebased commit. And if there were not enough reasons, fourth, git bisect works really well when you commits are grouped together.

For this issue I'd like us to agree to use squash+rebase when merging.  Second, I'd be interested to know if we can simplify our lives and make that the only rebase option available. I think this is a setting available to repo owners somewhere.

**Comment by ssoloff** [comment]
I'm all for squashing, as I believe each commit should ideally result in a green build.  We have a lot of PRs where there is follow-up work done before merging, and many of the intermediate commits represent broken builds.  These types of PRs should be squashed before merging, so, in that respect, I disagree with the rebase-only option.

On some occasions, a PR may contain readily identifiable discrete commits.  For example, a commit that adds a new feature, followed by a quasi-related refactoring.  It may make sense to not squash the commits in that case purely for history.  One could argue they should have been submitted in separate PRs, but sometimes it's just easier to push them together. :smile:

One thing that's bothered me when tracking down defect root causes is not always being able to map between commits and PRs.  Merge commits allow this because the PR number is stored in the merge commit message.  Squashing (by default) also retains this mapping by adding the PR number to the squashed commit message.  However, rebasing totally loses this information.

I don't think we always need to map between commits and PRs.  My philosophy has been to ignore the PR number and rebase if the PR is a refactoring or a minor bug fix.  For feature PRs, I'll either create a merge commit or squash with the PR number in the commit message so there is a back-reference to the PR in the history.  Maybe I'm just not aware of a GitHub feature that would allow one to easily map between commits and PRs. 😖

**Comment by RoiEXLab** [comment]
Difficult question.
I mean I try to group my commits logically so the message matches the purpose, and it's easy to understand what kind of changes were made, but I can see myself often pushing a commit that fixes a single line to make checkstyle happy or something. 
So I'm neutral on this, using merge commits where all of the commits are appended to the repo and a new merge commit is created to group those changes might be a compromise though.

**Comment by DanVanAtta** [comment]
> On some occasions, a PR may contain readily identifiable discrete commits. 

For review this is useful, but beyond I think the utility is lost. When reviewing git history I'd rather see one commit with: "implement feature XYZ with cleanup" rather than:
- prepping for feature
- adding feature
- feature fix part 1
- fix part 2
- cleanup
- PR feedback
- more refactoring

For a review the latter is useful, for a history it is a bit much. Considering two cases: 1 - mapping which PRs are in which release, we run into this issue:
![problem](https://user-images.githubusercontent.com/12397753/35197718-597471ae-fe98-11e7-83ca-ea546c1cae2b.png)

In the above only one PR was merged, yet we have many commits. To boot, if we rollback we'll likely revert the whole PR, it then makes it more difficult to know which commits were reverted when looking at the history (since it's not clear which commits are part of a merge, rebase helps but is still not perfect).

Second problem, when doing a git bisect or checking out different features for potential errors, the above has N commits to go through to find the problem, as compared to just the one if we were to squash.

> back-reference to the PR in the history

I sometimes think that is a bit of an anti-pattern, once a PR is closed I would suggest it has zero value, and negative value even if you must go back to look at it. Any such useful details need to ideally be captured in documentation or the commit message directly. To me it's a bit like the "Hey, I just forwarded this email to you, could you take care of it? [you open the email to see a thread that is 20 pages long with no direct mention to you or what is expected]"

> So I'm neutral on this, using merge commits where all of the commits are appended to the repo and a new merge commit is created to group those changes might be a compromise though.

That is basically rebase merge. I'm not sure why we need to compromise. Seeing the chain of commits is still difficult. IF you check out any one of those commits you get the state of that branch at that time, and not the overall code base. Rebasing really helps so the commits are grouped, but you are still having to track which N commits map to a merge commit, and again if you revert the merge commit it's not 100% clear which commits are going to be reverted. Last, frankly I don't want to see 20 commits in the history like "clean up checkstyle" followed by a merge commit (where anything can change), I would want to see it as one commit grouped by the high level update. Low level commit messages are useful during review, but after that they are clutter.


----------------------------------

to sum up, squash rebase has a number of benefits:
- one commit per PR or feature, this maps to what we roll back when we do
- one commit per feature makes it easy to debug the history
- history is cleaner, we have a 1:1 mapping between commits and reverts
- clarity of audience for commits. "PR feedback" is a horrible comment to live in the history on its own, but it is not bad in the scope of a PR pre-squash. Example, if tracking down a bug, you really are interested to know how an overall PR updated a block of code, going through all 20 incremental updates is not wanted (and we do not rebase, then those 20 commits would be scattered, so you'd have to track them down).

As is, it sounds like there are no supporters of just a standard merge commit. Squash + rebase I don't think has any real drawbacks at the moment, we get the benefits of all worlds I think with easier review and then valueable long term history. Let me know what you guys think, we should try to come to a concensus on this.

**Comment by DanVanAtta** [comment]
@ron-murhammer do you have any objections if we do only squash merges? Any thoughts on your side? I would ideally like to see that as the only merge option so that we have a consistent merge strategy.

**Comment by DanVanAtta** [comment]
Edit, I should make a cleaner case for squashes..

> I'm all for squashing

Excellent. I'm on board. 

> So I'm neutral on this, using merge commits where all of the commits are appended to the repo and a new merge commit is created to group those changes might be a compromise though.

@RoiEXLab the appended with merge is rebase + merge. 

--------------

Is there any case for simple merge commit?  Rebase or squash both seem nicer so that commits are grouped by branch/feature rather than merged by time of commit (which means branch can be 'retroactively' updated with a commit that takes effect a few days or months ago - it's hard to reason about branch state in this case).

-----------

Is there a case for rebase when we can squash?  Squash I think has all the benefits of rebase but you get even more. After PR, I tend to really want to see commits one per feature or PR. Going back one commit in history basically is an undo to the last change, there is no counting number of commits that were in any given branch before it is all undone. Beyond that there are still a few more:
- no extra merge commit
- fewer commits listed in the release section on github releases
- no broken interstitial commits, any commit on master would pass tests

**Comment by ssoloff** [comment]
> Is there any case for simple merge commit?

Outside of GitHub-flow, one would typically use a merge commit to resolve conflicts, if necessary, when merging a feature branch back to `master`.  Since GitHub-flow doesn't even allow one to merge when conflicts exist, this scenario is moot.

The only other benefit I see of using a merge commit is that it captures the commits on the feature branch just as the author made them.  That means things like signatures are preserved.  When using rebase or squash, author signatures are lost because GitHub rewrites the commits.  However, I'm not sure that's a compelling reason.

**Comment by DanVanAtta** [comment]
> The only other benefit I see of using a merge commit is that it captures the commits on the feature branch just as the author made them. That means things like signatures are preserved. 

Preserving signature is a benefit, for us marginal though. I would agree it's not a compelling reason to use merges. Preserving commits as author originally made them post-merge I would chalk up as a con even, not a benefit. Pre-merge it is a good/okay - but post-merge we are generally more concerned about the feature level. Commits like "resolved PR feedback" is okay to see in a PR, but otherwise it is a *bad* thing to see in history.

**Comment by DanVanAtta** [comment]
@ron-murhammer @RoiEXLab do either of you have strong objections to using squash-rebase merge strategy by default? I'd like to ideally even make it the only merge option so we do not even need to think/worry about it, but that is perhaps a next step.

I do see maintenance benefits with cleaner history, no merge commits to ever revert, 1:1 mapping between PRs and commits, would make updating release notes mostly a copy/paste exercise.

**Comment by RoiEXLab** [comment]
No objections.

**Comment by ron-murhammer** [comment]
@DanVanAtta Not really following what you are proposing here. I believe there are 3 main PR merge options:
1. Merge
2. Squash & Merge
3. Rebase

The default is Merge which I believe we all do today (though technically all 3 are available). Are you proposing 2 or 3? Maybe the best thing here is for each of us to vote/rank preferences of the options?

**Comment by DanVanAtta** [comment]
I've been clicking 3 or 2 usually : )

I propose we make the default '2. Squash and Merge' (and preferably make it the only option)

1. Merge: is actively bad from a history+debugging point of view.

2. + 3. Solve a number of the problems of a chronological merge, so either of these options really helps. I don't know if it's a matter of opinion, but post-merge I benefit most from seeing one commit for the PR instead of many. It helps in the release page, release notes and for revert.

So I'm really suggesting as a floor is we try to remove 'merge' as an option, make squash+rebase a default. If there is really strong support for not squashing, I could be persuaded. Chronological merge is a bit harmful, so I'd like us to stop using it.

>  Maybe the best thing here is for each of us to vote/rank preferences of the options?

:+1: 

I'm also curious though if I'm really misguided to value the single commit post-PR-merge. Maybe it might be better to decide first if we want to eliminate the 'Merge' option. Then we can focus between 2 or 3, and which trade-offs we would value more.

**Comment by DanVanAtta** [comment]
@ron-murhammer summarizing, maybe let's start looking at eliminating merge? That is my main objective, I'm curious if possible, and secondly if there is broad support and agreement about why it's harmful.

**Comment by ssoloff** [comment]
@RoiEXLab and I have been mostly using (2) during the past week (with a few (3)'s sprinkled in).  I'm really liking (2) as it makes the PR mapping clear in the history while keeping the history flat.  I know the former was noted as an anti-pattern, but I still find it useful. :smile:

**Comment by DanVanAtta** [comment]
Chronological merge (1) is actively harmful - @ron-murhammer would you mind turning it off? There have been no objections raised against this, and no support for keeping a standard merge around. I'll summarize one last time why merge is harmful for this project:
- impacts if not ruins git bisect. This is major. Git bisect is an invaluable tool for finding regressions.
- git history does not make a lot of sense. Chronological can see scattered commits across multiple months for one thread
- checking out commits in sequence drastically alters code base. One commit you're on one branch that started months ago, the previous commit you're back on master from a week ago.

**Comment by ron-murhammer** [comment]
I've removed the "merge" option now. I still think we should try to come to a consensus on squash vs rebase. If we find them both useful in certain situations then having a small guide on when to use each is fine. Just want us to have consistency across the project as if half of us are always using squash and the other half are always using rebase then I think we still get a weird history that is hard to follow.

**Comment by ssoloff** [comment]
My preferences:

1. Enable squash only
1. Enable both squash and rebase

I don't think enabling rebase only is a good option unless PR authors are willing to self-squash, when necessary, before merging.

**Comment by RoiEXLab** [comment]
I don't have any strong preferences on any method, although I like the squash merging on github, simply because the PR number is then included in the commit message

**Comment by DanVanAtta** [comment]
Even with 1, 2, 3 or 50 commits, I think if we see more value in one merge strategy over the other, we would always make the same decision. It probably makes sense to turn one or the other off.

I suspect it comes down to whether we will have, post-merge, well structured commits, or if we'll have things like:
A) added feature
B) PR feedback
C) fix tests
D) checkstyle
E) more feedback
F) merge

I suspect we'll have unstructured commits. I think the debate is settled for squash when we also consider some unique benefits to squash, namely that every commit on master would build. Second important benefit unique to squash is that we'll have fewer commits per release, each will map to a PR, it'll be easier to write release notes and know what is in each release:
![screenshot from 2018-02-08 19-04-48](https://user-images.githubusercontent.com/12397753/36010610-252c1d52-0d07-11e8-8780-c54c193b1ae3.png)

**Comment by DanVanAtta** [comment]
Shall we take a vote?

:tada: for squash only
:+1: for rebase only
:confused: allow both

**Comment by ssoloff** [comment]
:tada:

**Comment by RoiEXLab** [comment]
😕

**Comment by DanVanAtta** [comment]
Any strong feelings or particular reasons for both @RoiEXLab ?  Just curious. For me it is not a feature to leave it up to the merger, it's one more thing to think about, one more question that does not necessarily need to be answered every time we do a merge : )

**Comment by DanVanAtta** [comment]
@ron-murhammer any strong feelings towards one or the other, or happy with a consensus option?

**Comment by DanVanAtta** [comment]
Thanks, I think that means we are all leaning towards squash only or are neutral. That would seem to decide it for squash only.  @ron-murhammer, if no objection, could you make it so and remove the rebase option?

**Comment by ron-murhammer** [comment]
Disable rebase option since it seems we are in agreement on squash. Closing this issue now.
